third_party/move/move-vm/runtime/src/execution_tracing/loggers.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Interfaces and implementations for [Trace] collection.
+
+use crate::{
+    execution_tracing::{
+        trace::{CondBrTrace, DynamicCall},
+        Trace,
+    },
+    LoadedFunction,
+};
+use move_core_types::function::ClosureMask;
+
+/// Interface for recording the trace at runtime. It is sufficient to record branch decisions as
+/// well as dynamic function calls originating from closures.
+pub trait TraceLogger {
+    /// Returns true if the trace is being collected.
+    fn is_enabled(&self) -> bool;
+
+    /// Called in the end of execution to produce a final trace, suitable for replay.
+    fn finish(self) -> Trace;
+
+    /// Called after successful execution of a bytecode instruction. It is crucial that the trace
+    /// records only successful instructions.
+    fn record_successful_instruction(&mut self);
+
+    /// Called for every successfully executed conditional branch.
+    fn record_branch(&mut self, taken: bool);
+
+    /// Called for every successful set-up of the entrypoint (entry function or script). That is,
+    /// setting up frame, stack, and other structures before actually executing the bytecode.
+    fn record_entrypoint(&mut self, function: &LoadedFunction);
+
+    /// Called for every successful set-up of the closure call (i.e., immediately before the first
+    /// instruction of the callee is executed).
+    fn record_call_closure(&mut self, function: &LoadedFunction, mask: ClosureMask);
+}
+
+/// Logger that collects the full trace of execution. Records the number of successfully executed
+/// instructions, branch outcomes and closure calls.
+pub struct FullTraceLogger {
+    /// Number of successfully executed instructions.
+    ticks: u64,
+    /// Branch outcomes.
+    branches: CondBrTrace,
+    /// Dynamic call outcomes.
+    calls: Vec<DynamicCall>,
+}
+
+impl FullTraceLogger {
+    /// Returns a new empty logger ready for trace collection.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            ticks: 0,
+            branches: CondBrTrace::with_capacity(64),
+            calls: vec![],
```

Comment:
what's the typical size vs default capacity? maybe it can be worth it to also do with_capacity

Reply:
Good question! I think let's log these, and then decide? Added a few counters